{
  "name": "nerr",
  "version": "0.0.1",
  "description": "Properly constructed base class for error objects",
  "keywords": [
    "error",
    "exception"
  ],
  "author": {
    "name": "Dmitry Smolin",
    "email": "dimsmol@gmail.com"
  },
  "preferGlobal": false,
  "private": false,
  "engines": {
    "node": "~0.8.0"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/dimsmol/nerr"
  },
  "directories": {
    "lib": "./lib"
  },
  "main": "./lib/index",
  "readme": "# nerr\n\nUsing custom errors is kinda tricky in JS. Unfortunately, it is not enough to just inherit your class from Error. You need to do really strange things to provide stack trace with your error objects.\n\n**nerr** fixes this problem by providing ErrorBase class. By inheriting it you're getting correct error implementation with stack trace included.\n\nErrorBase features:\n\n* Inherited from Error, all your error objects will be instances of Error\n* Fully Error-compatible, mimics all it's properties, code working with Error will also work with classes inherited from ErrorBase\n* `message` is getter, not just string property - override getMessage() to construct message you need using any of your error's properties\n\n## How to use\n\n* Inherit from ErrorBase\n* Don't forget to call superclass constructor in your constructor\n* Override prototype.name by setting it to your class name\n* Override getMessage() to provide error message\n\nNote, that ErrorBase hasn't `message` argument that Error has. However, you can provide it in your error class yourself, as well as any other arguments and properties you may need. But be careful with property names. For example, inaccurate overriding of `message` property (which is a getter) can break functionality.\n\n## Usage example\n\nDefining custom error class:\n\n```js\nvar inherits = require('util').inherits;\nvar ErrorBase = require('nerr').ErrorBase;\n\nvar MyError = function (code) {\n\tErrorBase.call(this);\n\tthis.code = code;\n};\ninherits(MyError, ErrorBase);\n\nMyError.prototype.name = 'MyError';\n\nMyError.prototype.getMessage = function () {\n\treturn 'Error code: ' + this.code;\n};\n```\n\nUsing it:\n\n```js\ntry {\n\tthrow new MyError(500);\n}\ncatch(err) {\n\t// prints error name, message and stack trace\n\tconsole.log('Stack trace:\\n', err.stack);\n\n\tconsole.log();\n\n\t// prints error name and message\n\tconsole.log('Error:', err.toString());\n\n\tconsole.log();\n\n\tconsole.log('Error name:', err.name);\n\tconsole.log('Error message:', err.message);\n}\n```\n\n## How it works\n\nErrorBase constructor calls `Error.captureStackTrace()` which provides given object with a `stack` getter, that returns constructed stack trace.\n\n`captureStackTrace(obj, func)` constructs stack trace by concatenating obj.toString() with a stack trace itself (`func` is used to appropriately truncate stack trace, excluding error construction function trace). captureStackTrace() is called at construction time, when it isn't known what will be error's string representation - because it depends on getMessage() that may depend on properties that are not initialized yet. So, captureStackTrace() is provided with StackKeeper instance which has toString() returning empty string. Error's actual string representation will be added later, on call of getStackTrace() or `stack` getter.\n\n## Compatibility\n\nErrorBase implementation depends on undocumented `Error.captureStackTrace()` function and on how exactly it constructs a stack trace. For time of writing it's fully V8-compatible, but situation may change in future.\n\nIt's planned to provide fixes when needed, keeping ErrorBase backwards compatible. Please tell me if something is wrong with it on some specific version of node.js\n\n`require('nerr').compat.ErrorBase` contains implementation that does not depend on any undocumented features and supposed to work correctly on any version of node.js It produces less accurate stack trace with some lines that must not be actually present, but it is still a correct working solution. However, it is not recommended to use unless you need a temporary workaround, waiting for fix in standard ErrorBase.\n",
  "_id": "nerr@0.0.1",
  "dist": {
    "shasum": "7a03e8c438c1130881978a4923a31099fca5598f"
  },
  "_from": "nerr@~0.0.1"
}
